### PR TITLE
chore: Add core node status headers to gateway APIs

### LINKED ISSUE #5971: [data-node] [fairground] Data nodes serve stale data if their Vega node dies
# Ticket update
The original ticket suggested this was inconsistent data, which is *true*. However it's inconsistent because *n06 and n11 died*, and the data-node is serving old data from before the node died. Now the purpose of this ticket is how to represent that to the user so that they know not to trust the data they're seeing.

# Original ticket
I deposited some tDAI: https://ropsten.etherscan.io/tx/0x3aa570fa38a09e788835263b5942c4482a5708e5d94e69a0bd09da4d93f11192

My UI never successfully saw the deposit. So I went to Grafana and saw that the deposit had been seen:
image

So I went to GraphQL on n11 (the one the fairground console connects to by default) and ran:

```graphql
{
  party(id: "6eed3515524b3c6dda7a9eceaebba58e236a56faa8ef7b8129179c9d269f41f2") {
    depositsConnection {
      edges {
        node {
          asset { id name  }
          amount
          status
        }
      }
    }
  }
}
```

And got:

```json
{
  "data": {
    "party": {
      "depositsConnection": {
        "edges": null
      }
    }
  }
}
```

But how can that be? So I checked n10 instead, and got:

```json
{
  "data": {
    "party": {
      "depositsConnection": {
        "edges": [
          {
            "node": {
              "asset": {
                "id": "6d9d35f657589e40ddfb448b7ad4a7463b66efb307527fedd2aa7df1bbd5ea61",
                "name": "tDAI TEST"
              },
              "amount": "1000000000",
              "status": "STATUS_FINALIZED"
            }
          },
          {
            "node": {
              "asset": {
                "id": "5cfa87844724df6069b94e4c8a6f03af21907d7bc251593d08e4251043ee9f7c",
                "name": "tBTC TEST"
              },
              "amount": "11100000",
              "status": "STATUS_FINALIZED"
            }
          },
          {
            "node": {
              "asset": {
                "id": "5cfa87844724df6069b94e4c8a6f03af21907d7bc251593d08e4251043ee9f7c",
                "name": "tBTC TEST"
              },
              "amount": "12300000",
              "status": "STATUS_FINALIZED"
            }
          }
        ]
      }
    }
  }
}
```

A more complete survey showed that `n06-testnet` and `n11-testnet` return no TXs for my user, but `n07`, `n08`, `n09`, `n10`, `n12` return the transaction list.

# Solution
The Vega protocol design team decided the best fix for HTTP based clients is to use additional HTTP headers in the response to denote the state of the system. That way:
- We don't potentially break client implementations that rely on a `200` status code
- We don't have a bespoke REST or GraphQL endpoint that must be checked independently of the request you actually care about
- Existing implementations will be unaffected
- There is a clear indicator on HTTP responses

This doesn't address GRPC as the meeting didn't cover GRPC. If there's an easy adaptation, great.

# Task
- [ ] Every REST & GraphQL response should container the new headers:
  - [ ] `x-block-height`: '123'.
  - [ ] `x-block-timestamp: '1661338849'
- [ ] If a data node has not received blocks/events from its Node in a reasonable period of time, an additional header is set
  - [ ] `x-vega-connection: 'DISCONNECTED'` 

## Variations
- `x-block-timestamp` could also use the standard `Last modified` header if preferred. We went for block-timestamp because Last Modified could imply that the contents last changed then, which it may not have
- `x-vega-connection` could always be populated with a `CONNECTED` or `DISCONNECTED`
- `x-vega-connection` could have a clearer name

### PR DESCRIPTION
**Pull request by zale144**
Add all the changes necessary to add gRPC, GraphQL and REST APIs headers for communicating to the client about the state of the core node.

Closes #5971

### CODE REVIEW

**Review by pscott31 — DISMISSED** [review]
This looks good to me, though I admit I don't quite follow all the interceptor magic fully it seems sensible.

One thing that worries me a little bit is that by the looks of it this is going to add one grpc call to from datanode->core for every grpc call that happens on datanode. Some single GQL queries resolve into very many grpc calls - e.g. something `{trades{market{id}}` has a separate GRPC call to get the market for every trade.

Maybe GetState is so utterly trivial that it doesn't matter but its worth keeping an eye on, if it becomes a bother maybe we can fetch it once a second or something and cache it.

**Comment by zale144** [comment]
> This looks good to me, though I admit I don't quite follow all the interceptor magic fully it seems sensible.
> 
> One thing that worries me a little bit is that by the looks of it this is going to add one grpc call to from datanode->core for every grpc call that happens on datanode. Some single GQL queries resolve into very many grpc calls - e.g. something `{trades{market{id}}` has a separate GRPC call to get the market for every trade.
> 
> Maybe GetState is so utterly trivial that it doesn't matter but its worth keeping an eye on, if it becomes a bother maybe we can fetch it once a second or something and cache it.

`GetState` doesn't need any network calls or anything. It merely checks the value on a struct field `state` in 

```
type connectivityStateManager struct {
	mu         sync.Mutex
	state      connectivity.State
	notifyChan chan struct{}
	channelzID *channelz.Identifier
}
```

which is in the `google.golang.org/grpc` library. 

However `GetLastBlock` is a bit more expensive, it comes from the `BlockService` and ends up querying the db with this query: `SELECT vega_time, height, hash FROM blocks order by vega_time desc limit 1`

Like I have mentioned in the Slack thread, I still need to find a way to only call these methods once per request, because now if we use GQL, they get called double, because gRPC interceptors make one call to each method, and GQL middleware makes the another call, since the calls to the GQL are propagated to the gRPC client.

**Comment by jeremyletang** [comment]
> Like I have mentioned in the Slack thread, I still need to find a way to only call these methods once per request, because now if we use GQL, they get called double, because gRPC interceptors make one call to each method, and GQL middleware makes the another call, since the calls to the GQL are propagated to the gRPC client.

Makes sense, have you figured a way to do this? How about passing a flag on the grpc / rest call to indicate if you dont want the extra informations in the headers? This way you can just avoid all the calls made by the grpc handlers when going via graphql, and instead add it directly from the  graphql layer?

File: datanode/gateway/graphql/server.go [hunk 1]
```diff
@@ -23,15 +23,18 @@ import (
 	"strconv"
 	"time"
 
+	"golang.org/x/crypto/acme/autocert"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
```
Comment (jeremyletang) [inline datanode/gateway/graphql/server.go:28]: It's just style, but usually we try to have imports with the following order in the codebase:
- std import
- project import (code.vegaprotocol.io/*)
- external dependencies
Each block of imports being separated with a blank line.

File: datanode/gateway/middleware.go [hunk 2]
```diff
@@ -64,3 +65,55 @@ func MetricCollectionMiddleware(log *logging.Logger, next http.Handler) http.Han
...
+}
+
+type InjectableResponseWriter struct {
+	http.ResponseWriter
+	headers http.Header
+}
+
+type key string
+
+const injectableWriterKey key = "injectable-writer-key"
```
Comment (jeremyletang) [inline datanode/gateway/middleware.go:95]: We usually inject key like in the following file in the context: libs/context/context.go.

I also kinda seems to remember its the recommended way to do it.

**Comment by zale144** [comment]
> > Like I have mentioned in the Slack thread, I still need to find a way to only call these methods once per request, because now if we use GQL, they get called double, because gRPC interceptors make one call to each method, and GQL middleware makes the another call, since the calls to the GQL are propagated to the gRPC client.
> 
> Makes sense, have you figured a way to do this? How about passing a flag on the grpc / rest call to indicate if you dont want the extra informations in the headers? This way you can just avoid all the calls made by the grpc handlers when going via graphql, and instead add it directly from the graphql layer?

OK, I've pushed some changes. Basically, I'll be getting the headers from gRPC server, so no need to query the values from GQL API again

**Review by pscott31 — APPROVED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**